owid/catalog/catalogs.py

```diff
@@ -144,34 +145,43 @@ def _read_channels(self, channels: Iterable[CHANNEL]) -> pd.DataFrame:
             ),
         )
 
-    def iter_datasets(self, channel: CHANNEL) -> Iterator[Dataset]:
+    def iter_datasets(
+        self, channel: CHANNEL, include: Optional[str] = None
+    ) -> Iterator[Dataset]:
         to_search = [self.path / channel]
         if not to_search[0].exists():
             return
 
+        re_search = re.compile(include or "")
+
         while to_search:
             dir = heapq.heappop(to_search)
-            if (dir / "index.json").exists():
+            if (dir / "index.json").exists() and re_search.search(str(dir)):
                 yield Dataset(dir)
                 continue
 
             for child in dir.iterdir():
                 if child.is_dir():
                     heapq.heappush(to_search, child)
 
-    def reindex(self) -> None:
+    def reindex(self, include: Optional[str] = None) -> None:
```

Comment:
The logic seems correct, but honestly the `reindex()` method was already long and now it needs splitting. Could it look like this, at a high level?

```python
def reindex():
    index = _scan_for_datasets()
    _save_index(index)
    self.frame = index

def update_index(query):
    update = _scan_for_datasets(query)
    index = _merge_index(self.frame, update)
    _save_index(index)
    self.frame = index
```

Reply:
That's much better. Could you take a look now? I made some compromises but I'm not married to them.